expectation.messagematch.go

```diff
@@ -0,0 +1,192 @@
+package testkit
+
+import (
+	"errors"
+
+	"github.com/dogmatiq/configkit/message"
+	"github.com/dogmatiq/dogma"
+	"github.com/dogmatiq/testkit/envelope"
+	"github.com/dogmatiq/testkit/fact"
+	"github.com/dogmatiq/testkit/internal/inflect"
+)
+
+// ToExecuteCommandMatching returns an expectation that passes if a command is
+// executed that satisfies the given predicate function.
+//
+// Always prefer using ToExecuteCommand() instead, if possible, as it provides
+// more meaningful information in the result of a failure.
+//
+// desc is a human-readable description of the expectation. It should be phrased
+// as an imperative statement, such as "debit the customer".
+//
+// pred is the predicate function. It is called for each executed command. It
+// must return nil at least one for the expectation to pass.
+func ToExecuteCommandMatching(
+	pred func(dogma.Message) error,
+) Expectation {
+	if pred == nil {
+		panic("ToExecuteCommandMatching(<nil>): function must not be nil")
+	}
+
+	return &messageMatchExpectation{
+		pred:         pred,
+		expectedRole: message.CommandRole,
+	}
+}
+
+// IgnoreMessage is an error that can be returned by predicate functions used
+// with ToExecuteCommandMatching() and ToRecordEventMatching() to indicate that
+// the predicate does not care about the message and therefore the predicate's
+// result should not affect the expectation's result.
+var IgnoreMessage = errors.New("this message does not need to be inspected by the predicate")
+
+// messageMatchExpectation is an Expectation that checks that a message that
+// satisfies a predicate function is produced.
```

Comment:
Just check that this is clear. Each time I read it I feel like I need to read it again?

Reply:
extra `that`.